Migrate to Vue 3!

kilobyte2007 opened the pull request:
I have taken the time and migrated the project to Vue 3. The functionality is mostly intact as only dependency upgrades and the necessary migration refactoring were performed. Here's what was done:
1. Migrated all of the Vue code based on https://v3-migration.vuejs.org/
2. Switched the dev server to use vite
3. Upgraded webpack from 2 to 5
4. Upgraded several other dependencies
5. Restructured the build configurations to work with webpack 5
6. Adjusted the documentation to reflect changes

Please let me know if I missed something. I'd be happy if this PR was merged so I could move our own project to Vue 3.

Besides the above, I have several other bugfixes I'd like to discuss, but that's probably for a different thread.

kilobyte2007 committed:
Migration tu Vue 3

kilobyte2007 committed:
Cleanup

kilobyte2007 committed:
Fix Navbar emits

JosephusPaye commented on README.md:
````diff
@@ -51,62 +51,28 @@ You can add the reset to your stylesheet (before other styles). If you are using
 
 ```js
-import Vue from 'vue';
+import { createVue } from 'vue';
````
```suggestion
import { createApp } from 'vue';
```

JosephusPaye commented on README.md:
````diff
@@ -51,62 +51,28 @@ You can add the reset to your stylesheet (before other styles). If you are using
 });
 ```
 
-### Script tag
````
It would be good if we retain instructions on how use Keen UI via a `<script>` tag if possible. It's very useful when creating demos or prototyping in playgrounds like Codepen.

kilobyte2007 replied:
Ok, makes sense indeed but the way it is done now has changed, I've reflected that in the docs, so please le met know if it looks all right.

JosephusPaye commented on README.md:
````diff
@@ -120,20 +86,20 @@ Each component is built into a standalone file with the necessary CSS included.
 ```
 
-## Licence
+## License
 
-Keen UI is open source and released under the [MIT Licence](LICENCE).
````
"Licence" is the correct spelling in the [Queen's English](https://www.grammarly.com/blog/licence-license/#:~:text=License%20as%20a%20Noun%3A%20Spelling%20and%20Examples&text=In%20American%20English%2C%20the%20noun,or%20allowed%20to%20do%20something.) 😅

kilobyte2007 replied:
All right, it was just my IDE bugging me :-D

JosephusPaye commented on build/webpack.dist.js:
```diff
@@ -18,46 +17,15 @@ const config = merge(base, {
-        })
-    ]);
-}
+}, true);
```
We could make this take an options object so the option name provides context for what the boolean is without the need to check the function definition. Same applies for the other webpack config files.

```suggestion
}, { extractCss: true });
```

kilobyte2007 replied:
Sounds good.

JosephusPaye commented on build/webpack.dist.js:
```diff
@@ -1,13 +1,12 @@
 const base = require('./webpack.base.js');
 
-const config = merge(base, {
+const config = base({
```
I like the refactor to use a function instead of a deep merge here 👍

JosephusPaye commented on docs-src/index.html:
```diff
@@ -0,0 +1,13 @@
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
+    <title>Keen-UI</title>
```
```suggestion
    <title>Keen UI</title>
```

JosephusPaye commented on docs-src/router/index.js:
```diff
@@ -0,0 +1,245 @@
+    };
+});
+
+const router = createRouter({
```
Needed for the active link indicator in the sidebar.

```suggestion
const router = createRouter({
    linkActiveClass: 'is-active',
```

JosephusPaye commented on docs-src/router/index.js:
```diff
@@ -0,0 +1,245 @@
+import { createRouter, createWebHashHistory } from 'vue-router'
```
We should probably keep this file in the `docs-src` folder, since it's the only one here, it's not necessary to add a `router/` subfolder.

I like the name `navigation.js` too, since this exports not only a router, but the menu items as well. I'm not too fussed about the name though, and can go with `router.js`.

kilobyte2007 replied:
Seems sensible.

JosephusPaye commented on docs-src/router/index.js:
```diff
@@ -0,0 +1,245 @@
+        },
+        ...routes,
+        {
+            path: '/.*',
```
```suggestion
            // Catch all route to redirect 404s
            path: '/.*',
```

JosephusPaye commented on docs-src/pages/UiTextbox.vue:
```diff
@@ -225,7 +225,7 @@
                                 <td>
                                     <p>The model that the textbox value syncs to. Can be set initially as a default value.</p>
-                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>value</code>.</p>
+                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>modelValue</code>.</p>
```
```suggestion
                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>update:modelValue</code> event and update <code>modelValue</code>.</p>
```

JosephusPaye commented on docs-src/pages/UiTextbox.vue:
```diff
@@ -225,7 +225,7 @@
                                 <td></td>
```
Can't comment on the unchanged line, but the `value` prop above should now be `modelValue`.

JosephusPaye commented on docs-src/pages/UiTextbox.vue:
```diff
@@ -537,11 +537,11 @@
-                                <td>input</td>
+                                <td>update:modelValue</td>
                                 <td>
                                     <p>Emitted when the select value is changed. The handler is called with the new value.</p>
```
```suggestion
                                    <p>Emitted when the textbox value is changed. The handler is called with the new value.</p>
```

JosephusPaye commented on src/UiTabs.vue:
```diff
@@ -13,19 +13,16 @@
-                        v-else-if="hasIcon && Boolean(tab.$slots.icon)"
-                    ></render>
+                    <render :nodes="tab.$slots.header()" v-if="tab.$slots.header"></render>
+                    <template v-if="!tab.$slots.header && hasIcon && Boolean(tab.$slots.icon)" #icon>
```
Any reason to go from `v-else-if` to `v-if` with a negation here? Just curious.

kilobyte2007 replied:
Yes, in Vue 3 you cannot use v-else-if on a slot so it throws an error: ` v-else/v-else-if has no adjacent v-if or v-else-if.`

I have read about it on the Vue forums, and it seems it is the intended behavior.

JosephusPaye commented on docs-src/pages/UiSwitch.vue:
```diff
@@ -81,7 +81,7 @@
                                 <td>
```
Can't comment on unchanged line above, but `value` prop should now be `modelValue`.

JosephusPaye commented on docs-src/pages/UiSwitch.vue:
```diff
@@ -81,7 +81,7 @@
                                     <p>The model that the switch value syncs to.</p>
                                     <p>The <code>trueValue</code> prop will be written to this model when the switch is turned on and the <code>falseValue</code> prop will be written to it when the switch is turned off.</p>
-                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>value</code>.</p>
+                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>modelValue</code>.</p>
```
```suggestion
                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>update:modelValue</code> event and update <code>modelValue</code>.</p>
```

JosephusPaye commented on docs-src/pages/UiSlider.vue:
```diff
@@ -80,7 +80,7 @@
                                 <td>
                                     <p>The model that the slider value syncs to. Changing this value will update the slider.</p>
-                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>value</code>.</p>
+                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>modelValue</code>.</p>
```
```suggestion
                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>update:modelValue</code> event and update <code>modelValue</code>.</p>
```

JosephusPaye commented on docs-src/pages/UiSlider.vue:
```diff
@@ -80,7 +80,7 @@
                                 <td>(required)</td>
```
Can't comment on unchanged line above, but the `value` prop should now be `modelValue`.

JosephusPaye commented on docs-src/pages/UiSelect.vue:
```diff
@@ -230,7 +230,7 @@
                                 <td>
                                     <p>The model the selected value or values sync to. Can be set initially for default value/values.</p>
-                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>value</code>.</p>
+                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>modelValue</code>.</p>
```
```suggestion
                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>update:modelValue</code> event and update <code>modelValue</code>.</p>
```

JosephusPaye commented on docs-src/pages/UiSelect.vue:
```diff
@@ -230,7 +230,7 @@
                                 <td></td>
```
Can't comment on unchanged line above, but the `value` prop should now be `modelValue`.

JosephusPaye commented on docs-src/pages/UiRadioGroup.vue:
```diff
@@ -115,7 +115,7 @@
                                 <td>
                                     <p>The model that the selected value in the radio group syncs to. Can be set initially for a default selection.</p>
-                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>value</code>.</p>
+                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>modelValue</code>.</p>
```
```suggestion
                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>update:modelValue</code> event and update <code>modelValue</code>.</p>
```

JosephusPaye commented on docs-src/pages/UiRadioGroup.vue:
```diff
@@ -115,7 +115,7 @@
                                 <td></td>
```
Can't comment on unchanged line above, but the `value` prop should now be `modelValue`.

JosephusPaye commented on docs-src/pages/UiRadio.vue:
```diff
@@ -81,7 +81,7 @@
                                 <td>
                                     <p>The model the radio value syncs to.</p>
-                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>value</code>.</p>
+                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>modelValue</code>.</p>
```
```suggestion
                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>update:modelValue</code> event and update <code>modelValue</code>.</p>
```

JosephusPaye commented on docs-src/pages/UiRadio.vue:
```diff
@@ -81,7 +81,7 @@
                                 <td>Required</td>
```
Can't comment on unchanged line above, but the `value` prop should now be `modelValue`.

JosephusPaye commented on src/UiMenu.vue:
```diff
@@ -21,9 +21,9 @@
-            @keydown.enter.native="selectOption(option)"
-            @keydown.esc.native.esc="closeMenu"
+            @click="selectOption(option)"
+            @keydown="selectOption(option)"
```
```suggestion
            @keydown.enter="selectOption(option)"
```

JosephusPaye commented on src/UiFileupload.vue:
```diff
@@ -13,7 +13,7 @@
 
             @blur="onBlur"
-            @input="onInput"
+            @update:modelValue="onInput"
```
This component doesn't support v-model, and instead just re-emits the `input` event from the file input. So we don't need to change these: keep the `input` event.

kilobyte2007 replied:
Yes, just a mechanical error.

JosephusPaye commented on docs-src/pages/UiDatepicker.vue:
```diff
@@ -232,7 +232,7 @@
                                 <td>
                                     <p>The model the selected date syncs to. Can be set initially for a default value.</p>
-                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>value</code>.</p>
+                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>modelValue</code>.</p>
```
```suggestion
                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>update:modelValue</code> event and update <code>modelValue</code>.</p>
```

JosephusPaye commented on docs-src/pages/UiDatepicker.vue:
```diff
@@ -232,7 +232,7 @@
                                 <td></td>
```
Can't comment on unchanged line above, but the `value` prop should now be `modelValue`.

JosephusPaye commented on docs-src/pages/UiCheckboxGroup.vue:
```diff
@@ -103,7 +103,7 @@
                                 <td>
                                     <p>The model that holds the checked option values. Can be set initially for default values.</p>
-                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>value</code>.</p>
+                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>modelValue</code>.</p>
```
```suggestion
                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>update:modelValue</code> event and update <code>modelValue</code>.</p>
```

JosephusPaye commented on docs-src/pages/UiCheckboxGroup.vue:
```diff
@@ -103,7 +103,7 @@
                                 <td></td>
```
Can't comment on unchanged line above, but the `value` prop should now be `modelValue`.

JosephusPaye commented on src/UiCheckbox.vue:
```diff
@@ -31,11 +31,13 @@
 export default {
     name: 'ui-checkbox',
 
+    emits: ['update:modelValue', 'changes', 'focus', 'blur', 'change'],
```
```suggestion
    emits: ['update:modelValue', 'focus', 'blur', 'change'],
```

JosephusPaye commented on docs-src/pages/UiCheckbox.vue:
```diff
@@ -89,7 +89,7 @@
                                     <p>The model that the checkbox value syncs to.</p>
                                     <p>The <code>trueValue</code> prop will be written to this model when the checkbox is checked and the <code>falseValue</code> prop will be written to it when the checkbox is unchecked.</p>
-                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>value</code>.</p>
+                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>modelValue</code>.</p>
```
```suggestion
                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>update:modelValue</code> event and update <code>modelValue</code>.</p>
```

JosephusPaye commented on docs-src/pages/UiCheckbox.vue:
```diff
@@ -89,7 +89,7 @@
                                 <td>
```
Can't comment on unchanged line above, but the `value` prop should now be `modelValue`.

JosephusPaye commented on docs-src/pages/UiCalendar.vue:
```diff
@@ -152,7 +154,7 @@
                                 <td>
                                     <p>The model the selected calendar date syncs to.</p>
-                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>value</code>.</p>
+                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>modelValue</code>.</p>
```
```suggestion
                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>update:modelValue</code> event and update <code>modelValue</code>.</p>
```

JosephusPaye commented on docs-src/pages/UiCalendar.vue:
```diff
@@ -152,7 +154,7 @@
                                 <td></td>
```
Can't comment on unchanged line above, but the `value` prop should now be `modelValue`.

JosephusPaye commented on docs-src/pages/UiAutocomplete.vue:
```diff
@@ -151,7 +151,7 @@
                                 <td>
                                     <p>The model the autocomplete value syncs to.</p>
-                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>value</code>.</p>
+                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>input</code> event and update <code>modelValue</code>.</p>
```
```suggestion
                                    <p>If you are not using <code>v-model</code>, you should listen for the <code>update:modelValue</code> event and update <code>modelValue</code>.</p>
```

JosephusPaye commented on docs-src/pages/UiAutocomplete.vue:
```diff
@@ -151,7 +151,7 @@
                                 <td></td>
```
Can't comment on unchanged line above, but the `value` prop should now be `modelValue`.

JosephusPaye reviewed (changes requested):
Thanks a lot for the PR. I can imagine it took a lot of time. ❤️

Some comments that couldn't be attached to a file:

- Revert "LICENCE" file rename - [it's the the right spelling](https://www.grammarly.com/blog/licence-license/#:~:text=License%20as%20a%20Noun%3A%20Spelling%20and%20Examples&text=In%20American%20English%2C%20the%20noun,or%20allowed%20to%20do%20something.) (if you're not American 😛)
- If possible, it would be good to keep the behaviour that allows for Vue Devtools inspection in production on the docs, as it's a useful learning tool. Perhaps this will help: https://github.com/vuejs/devtools/issues/1321
- The in transition for the sidebar on mobile seems to have broken in the docs. It appears instantly instead of sliding in. The out transition works though.

---

<details>

<summary>Other thoughts beyond this PR, left here for future reference</summary>

- We should look into bundling the library and docs with Vite instead of Webpack, so we have one less dependency. For now we can fix the warnings on the docs production build.

<details>
<summary>View warnings</summary>

> runtime-core.esm-bundler.js:5075 Feature flags **VUE_OPTIONS_API**, **VUE_PROD_DEVTOOLS** are not explicitly defined. You are running the esm-bundler build of Vue, which expects these compile-time feature flags to be globally injected via the bundler config in order to get better tree-shaking in the production bundle.
> For more details, see https://link.vuejs.org/feature-flags.

> DevTools failed to load source map: Could not load content for webpack://keen-ui/node_modules/tippy.js/esm/index.js.map: Fetch through target failed: Unsupported URL scheme; Fallback: HTTP error: status code 404, net::ERR_UNKNOWN_URL_SCHEME

This might help: https://stackoverflow.com/questions/61767538/devtools-failed-to-load-sourcemap-for-webpack-node-modules-js-map-http-e

</details>

- Breaking changes that will need documenting for the release:
  - Keen UI will no longer be installed automatically if Vue is available globally
  - `v-model/value` to `v-model/modelValue` changes
  - others
- UiSelect "With validation" example doesn't seem to be validating (not this PR's fault, also broken in the current docs)
- The popover in `UiIconButton.vue` has a typo for the `contain-focus` prop - it currently has `constain-focus`
- ImagePane component on the UiRippleInk docs page has an error:
  > Component provided template option but runtime compilation is not supported in this build of Vue. Configure your bundler to alias "vue" to "vue/dist/vue.esm-bundler.js"
- UiDatepickerCalendar component has a warning about accessing a variable that's not defined:

  > Property "year" was accessed during render but is not defined on instance.

  Issue is `this.year` in the isYearOutOfRange() method. That should be just `year`.

- The year list `v-for` in UiDatepickerCalendar should have a key, `year`.
- The `v-if` on the year list iterator in UiDatepickerCalendar should be removed, with a new computed added that can be used for the `v-for`:

  ```js
  yearRangeFiltered() {
    return this.yearRange.filter(year => !this.isYearOutOfRange(year));
  }
  ```

- UiCloseButton has a warning about accessing a variable that's not defined:

  > Property "loading" was accessed during render but is not defined on instance.

  Need to remove `this.loading` from the `classes` computed.

</details>

kilobyte2007 committed:
PR comments and fixes

kilobyte2007 committed:
PR comments - small bugfixes

kilobyte2007 commented:
Hey @JosephusPaye,
Thanks a lot for the detailed commentaries, I have fixed almost all of the issues as well as the bugs and warnings you mentioned.

1. As for using vite for all the builds, I agree with that, we'll have to look into this, as this might make more sense, I am just not that familiar yet with vite and rollup to configure it as confidently as I can configure webpack.
2. The devtool still works even without the flag when running the production build of docs.
3. As for the breaking changes - yes, would it be better to document them in the release notes/changelog somewhere?
4. Also I added a fix for the calendar as it was inheriting some global styles for `table` from the docs but it was looking differently when using KeenUI in our project as that style was not included in the bundle. Maybe it makes sense to never use global `table` and `td` styles in the docs CSS so that we can be sure that the docs reflect everything correctly.
5. The UiSelect issue I managed to fix as it seems that both the ui-popover and the ui-select were listening to the outside click event, it seems to work perfectly now, but please take a look too yourself.

All the other bugs and comments I have fixed and it should work well now. Please take a look and let me know what you think.

kilobyte2007 committed:
PR comments - small bugfixes and improvements

kilobyte2007 committed:
License -> Licence revert

kilobyte2007 commented:
Another thing I think we should do in this release is some housekeeping - updating the packages where we can, as some of them throw deprecation warnings and have security vulnersbilities:
1. tippy to 6
2. deepmerge to 4
3. autosize to 4
4. autoprefixer to 10
5. cssloader to 6
6. eslint to 8 (same for all eslint packages)
7. rimraf to 3

tippy might have some breaking changes but others should not require any code changes. What do you think?

JosephusPaye commented:
Thanks for making the changes!

> The devtool still works even without the flag when running the production build of docs.

That's good to know.

> As for the breaking changes - yes, would it be better to document them in the release notes/changelog somewhere?

Yeah, [CHANGELOG.md](https://github.com/JosephusPaye/Keen-UI/blob/next/CHANGELOG.md) is where we should document the breaking changes.

> Also I added a fix for the calendar as it was inheriting some global styles for table from the docs but it was looking differently when using KeenUI in our project as that style was not included in the bundle. Maybe it makes sense to never use global table and td styles in the docs CSS so that we can be sure that the docs reflect everything correctly.

Agreed. We can add a class to apply the docs-specific table styles, and only use it where necessary. You change to the component looks good though, we can keep that.

> The UiSelect issue I managed to fix as it seems that both the ui-popover and the ui-select were listening to the outside click event, it seems to work perfectly now, but please take a look too yourself.

Could you say more what the issue was?

> Another thing I think we should do in this release is some housekeeping - updating the packages where we can, as some of them throw deprecation warnings and have security vulnersbilities:

Agreed. Though it would be better to that in a separate PR. The Tippy breaking change should be OK if we document it along with the other breaking changes.

kilobyte2007 committed:
Fix broken placeholders in docs

kilobyte2007 commented:
As for the UiSelect issue - the thing was that both the UiPopover and the UiSelect were listening to an outside click, so when the outside click event was caught in UiSelect - the UiPopover's `isOpen` returned false as it was already closed by that point so the `closeDropdown` method was never called inside UiSelect which ultimatelly emits a `touch` event that was used in the validation example. What I had to do is to disable `close-on-external-click` for the UiPopover inside the UiSelect as this behavior is controlled at the UiSelect level.

I have also gone ahead and fixed the broken image placeholders (placehold.it moved to via.placeholder.com now).

As for the tables - I think we can just revert to using only `.table` instead of the `table` in the docs CSS files so that we don't mix it with the component's actual styles because wherever a table is used the class is present now. Should I include this change in this PR too?

As for the Changelog, should I do this in this PR or will you document that when you release a new version? Here's a breakdown of the breaking changes and bugfixes:

## Breaking changes
1. `modelValue` should be used on all components where the `value` prop was used. This includes the following components: UiTextbox, UiSelect, UiCalendar, UiCheckbox, UiCheckboxGroup, UiRadioGroup, UiDatepicker, UiDatepickerCalendar, UiAutocomplete, UiSlider, UiSwitch.
2. The `update:modelValue` event should be used wherever the `input` event was used. This includes the following components: UiTextbox, UiSelect, UiCalendar, UiCheckbox, UiCheckboxGroup, UiRadioGroup, UiDatepicker, UiDatepickerCalendar, UiAutocomplete, UiSlider, UiSwitch, UiFileupload.
3. The way KeenUI can be used globally has changed.

## Fixes and updates
1. The `touch` event was fixed for the UiSelect component. It was not fired previously when clicking away from the dropdown.
2. The UiCalendar styling was fixed as it was inheriting some styles from the docs site.
4. Fixed the UiIconButton focus containing when used with a dropdown.
5. UiCloseButton - fixed warning (this.loading check was not needed here)
6. UiDatepickerCalendar - fixed usage of a year range.

Please let me know what else is needed for me in order for this PR to be merged. Thanks!

kilobyte2007 committed:
Revert to Licence

kilobyte2007 committed:
Prevent table css styles from overlapping with UI element styles (e.g. UiCalendar).

kilobyte2007 commented:
Another thing. I went ahead a made a small improvement to some components that emit a new value if a previous one was invalid. Previously there were possibilities of events being emitted even though the value has never changed actually (UiCheckbox, UiSwitch) or in the case of UiSelect - it was always emitting an `input` and a `change` event even though the value was equal `''`.

Please take a look and let me know what you think of it, I'd be happy if we could add this change to this PR as well as it would help solve a lot of issues. Commit - https://github.com/kilobyte2007/Keen-UI/commit/047c211e294df003a211090e16a0645cc309d8a7

JosephusPaye commented on README.md:
```diff
@@ -51,38 +51,31 @@ You can add the reset to your stylesheet (before other styles). If you are using
-### Script tag
-
-First, add a stylesheet link to the Keen UI CSS file in `dist/keen-ui.min.css`. Then, add a script tag pointing to `dist/keen-ui.min.js` _**after**_ adding Vue.
+First, add a stylesheet link to the Keen UI CSS file in `dist/keen-ui.min.css`. Then, add a script tag pointing to `dist/keen-ui.min.js`.
```
```suggestion
### Script tag

First, add a stylesheet link to the Keen UI CSS file in `dist/keen-ui.min.css`. Then, add a script tag pointing to `dist/keen-ui.min.js`.
```

JosephusPaye commented on docs-src/styles/tables.scss:
```diff
@@ -1,36 +1,32 @@
-// From Bootstrap: http://getbootstrap.com/customize/?id=e516e5d6e28288745392
-table {
+caption {
```
We can prefix this with the `table` class as well, and probably move it to after the `.table` block.

```suggestion
.table caption {
```

kilobyte2007 replied:
Sure!

kilobyte2007 replied:
Done!

JosephusPaye commented on src/UiFileupload.vue:
```diff
@@ -48,6 +48,8 @@ import UiRippleInk from './UiRippleInk.vue';
 export default {
     name: 'ui-fileupload',
 
+    emits: ['update:modelValue', 'focus', 'blur', 'change'],
```
I think we can keep the event here as just `input`. Since it can't be used with `v-model`, there's no expectation for it to be `update:modelValue`.

```suggestion
    emits: ['input', 'focus', 'blur', 'change'],
```

kilobyte2007 replied:
Yes, been thinking about this, makes sense.

kilobyte2007 replied:
Done!

JosephusPaye commented on src/UiFileupload.vue:
```diff
@@ -139,7 +141,7 @@ export default {
 
         onInput(e) {
-            this.$emit('input', this.$refs.input.files, e);
+            this.$emit('update:modelValue', this.$refs.input.files, e);
```
```suggestion
            this.$emit('input', this.$refs.input.files, e);
```

kilobyte2007 replied:
Done!

JosephusPaye commented:
> As for the UiSelect issue - the thing was that both the UiPopover and the UiSelect were listening to an outside click, so when the outside click event was caught in UiSelect - the UiPopover's isOpen returned false as it was already closed by that point so the closeDropdown method was never called inside UiSelect which ultimatelly emits a touch event that was used in the validation example. What I had to do is to disable close-on-external-click for the UiPopover inside the UiSelect as this behavior is controlled at the UiSelect level.

Makes sense.

> I have also gone ahead and fixed the broken image placeholders (placehold.it moved to via.placeholder.com now).

Great, thanks!

> As for the tables - I think we can just revert to using only .table instead of the table in the docs CSS files so that we don't mix it with the component's actual styles because wherever a table is used the class is present now. Should I include this change in this PR too?

Yeah that's a good approach. I think what you've added now (the `.table` prefix) is OK for this PR.

> As for the Changelog, should I do this in this PR or will you document that when you release a new version? Here's a breakdown of the breaking changes and bugfixes:

Thanks for the list of changes. I'll add them to the changelog after we merge this PR.

> Please take a look and let me know what you think of it, I'd be happy if we could add this change to this PR as well as it would help solve a lot of issues. Commit - https://github.com/kilobyte2007/Keen-UI/commit/047c211e294df003a211090e16a0645cc309d8a7

Let's do this one in a separate PR.

> Please let me know what else is needed for me in order for this PR to be merged. Thanks!

This PR is very close to merging. I've done another review with your latest changes. Once you address the comments we can merge this!

Thanks a lot for your efforts on this.

kilobyte2007 committed:
Update README.md

Co-authored-by: Josephus Paye II <j.paye96@gmail.com>

kilobyte2007 committed:
PR comments

kilobyte2007 commented:
Hey @JosephusPaye, I just implemented the changes. Let me know if there's something else.
For the `emits` proposal - I will create a separate PR once this one is merged. Thanks for taking the time to review everything. When the change is so big it's easy to miss some things even when I try to be very careful :-)

kilobyte2007 committed:
Fix directive for Vue 3

kilobyte2007 committed:
Fix UiSelect slot verification

kilobyte2007 committed:
Fix transition classes

JosephusPaye reviewed (changes requested):
We're almost there, just a few more things I noticed:

- `beforeDestroy()` needs to be renamed to `beforeUnmount()` in the doc components App.vue, UiProgressCircular.vue and UiProgressLinear.vue

- The days `v-for` in UiCalendarMonth is missing a key.

- `v-if` and `v-for` on the same element for the snap points in UiSlider can be updated to have a template wrapping the span, with the `v-if` on the template, and the `v-for` remaining on the span.

kilobyte2007 committed:
PR Comments and fixes

kilobyte2007 commented:
Thanks a lot for catching those! Fixed now.

JosephusPaye reviewed (approved):
Looks good! Thanks again for working on this. Any other changes needed can follow in other PRs.

kilobyte2007 commented:
No problem at all, happy to move the library further and use it in our projects.
...and sure, I have submitted a few by now, and I will be submitting a few more, mostly very small ones. Let's discuss that in there :-)
